From c374e6aa606dd58026272e74c0d23aeedfb80891 Mon Sep 17 00:00:00 2001
From: Sophie 'Sharky' Schumann <sharky@sharky.pw>
Date: Thu, 25 Apr 2019 18:28:08 +0200
Subject: [PATCH] Fixes for recent Dovecot ABI and nullptr crashfix when using
 Dovecot-LDA

---
 Makefile      |  4 ++--
 drac-plugin.c | 27 +++++++++++++++++----------
 2 files changed, 19 insertions(+), 12 deletions(-)

diff --git a/Makefile b/Makefile
index 3c9725c..62b1f8e 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
-DOVDIR:=../dovecot-2.0.13
+DOVDIR:=/usr/include/dovecot/
 LIBS:=-ldrac
-LDFLAGS:=-L/usr/local/lib
+LDFLAGS:=-L/usr/lib
 
 CFLAGS=-Wall -W -shared -fPIC -DHAVE_CONFIG_H -I$(DOVDIR) \
 	-I$(DOVDIR)/src/lib \
diff --git a/drac-plugin.c b/drac-plugin.c
index ebe215c..2c50c6d 100644
--- a/drac-plugin.c
+++ b/drac-plugin.c
@@ -10,7 +10,7 @@
  *   http://dovecot.org/patches/1.1/drac.c
  */
 #include "lib.h"
-#include "network.h"
+#include "net.h"
 #include "ioloop.h"
 #include "mail-user.h"
 #include "mail-storage-private.h"
@@ -25,7 +25,7 @@
 /* libdrac function */
 int dracauth(char *, unsigned long, char **);
 
-const char *drac_plugin_version = DOVECOT_VERSION;
+const char *drac_plugin_version = DOVECOT_ABI_VERSION;
 
 static struct timeout *to_drac = NULL;
 static const char *drachost = NULL; /* dracd host */
@@ -43,22 +43,29 @@ static void drac_timeout(void *context ATTR_UNUSED)
 
 static void drac_mail_user_created(struct mail_user *user)
 {
+    i_debug("%s called", __FUNCTION__);
     const char *dractout_str;
-    char addrname[256];
     char *ep;
 
+    /*i_debug("%s:%d Checking address family", __FUNCTION__, __LINE__);
+    i_debug("user: %X", user);
+    if(user != NULL) i_debug("remote_ip: %X", user->remote_ip);
+    if(user->remote_ip != NULL) i_debug("family: %X", user->remote_ip->family);*/
+
+    if(user == NULL || user->remote_ip == NULL) {
+        i_error("%s:%d: no remote_ip found for user", __FUNCTION__, __LINE__);
+        return;
+    }
+
     /* check address family */
-    if(user->remote_ip->family != AF_INET) {
-        if(inet_ntop(user->remote_ip->family, &user->remote_ip->u,
-           addrname, sizeof(addrname)-1) == NULL) {
-            strcpy(addrname, "<unknown>");
-        }
-        i_error("%s: Only IPv4 addresses are supported (%s)", __FUNCTION__,
-                addrname);
+    if(user && user->remote_ip && user->remote_ip->family && user->remote_ip->family != AF_INET) {
+        i_error("%s: Only IPv4 addresses are supported", __FUNCTION__);
     } else {
         /* get remote IPv4 address... uum... */
         memcpy(&in_ip, &user->remote_ip->u, sizeof(in_ip));
 
+        i_debug("%s:%d Get IP: %lu", __FUNCTION__, __LINE__, in_ip);
+
         /* get DRAC server name */
         drachost = mail_user_plugin_getenv(user, "dracdserver");
         if(drachost == NULL) {
-- 
2.17.1.windows.2

